### PR TITLE
fix(local-ydb): fix root password rotation

### PR DESCRIPTION
**Pull request by astandrik**
<h3>Greptile Summary</h3>

This PR fixes several correctness and security issues in the root-password rotation flow: passwords containing backslashes were silently corrupted in the YQL query (only single-quotes were escaped, not backslashes), and the `rootUser` name was interpolated directly into the ALTER USER statement, making it injectable. The fixes pass the username as a data argument to Ruby and introduce a `yql_identifier` helper for safe backtick quoting.

- **Backslash escaping**: both the TypeScript `escapedPassword` (used for redaction) and the Ruby `sql_escaped` builder now double-escape `\` before escaping `'`, closing a silent data-corruption bug for any password containing a backslash.
- **rootUser injection fix**: `rootUser` is now passed as `ARGV.fetch(1)` to the Ruby script and wrapped with `yql_identifier`, which escapes `\` and `` ` `` before quoting; a matching test with a backtick-injected username verifies the fix.
- **Trap hardening**: all `trap` handlers now capture `$?`, extend to `HUP INT TERM`, clear themselves before exiting, and re-emit the original exit code, ensuring temp files are cleaned up on signal and that the caller sees the correct status.

<h3>Confidence Score: 5/5</h3>

Safe to merge — the changes are well-scoped fixes with corresponding tests and no regressions observed.

All changed code paths have matching test coverage, the injection fix is correctly implemented (rootUser passed as data, not interpolated into SQL), the backslash-escaping bug is fixed consistently in both TypeScript and Ruby, and the trap hardening is mechanically correct. No unguarded code paths or dropped state found.

No files require special attention.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| packages/core/src/operations/auth-operations.ts | Core auth operation fixes: proper backslash escaping in passwords, rootUser injection prevention via yql_identifier + argv data passing, improved trap signal handling with exit code propagation, and explicit UTF-8 file read. |
| packages/core/test/operations.test.ts | Tests expanded to verify trap ordering, signal handling, escape sequences, rootUser injection safety, and both CR and LF rejection — all aligned with the new implementation. |
| packages/mcp-server/src/tools/args.ts | Password validation switched from `.regex()` to `.refine()`, which is semantically equivalent here but clearer in intent; no functional regression. |
| packages/mcp-server/src/tools/input-schemas.ts | JSON Schema pattern updated to `^(?!.*[\r\n]).+$`, which is stricter than the old `^[^\r\n]+$` (the old pattern could match a string ending with `\n` in ECMAScript where `$` matches before a final newline). |
| packages/mcp-server/test/tools.test.ts | New tests cover the updated schema pattern (CR, LF rejection) and Zod validation path; straightforward and correct. |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant TS as TypeScript (setRootPassword)
    participant OB as Outer Bash (rotateSpec)
    participant Ruby as Ruby (query generator)
    participant IB as Inner Bash (rotate_with_password_file)
    participant YDB as /ydb CLI (container)
    participant SB as Sync Bash (syncHostSpec)
    participant RubySync as Ruby (config updater)

    TS->>OB: "stdin=newPassword"
    OB->>Ruby: "stdin=newPassword, ARGV[0]=$query_host, ARGV[1]=rootUser"
    Ruby->>Ruby: yql_identifier(rootUser) → backtick-quoted identifier
    Ruby->>Ruby: "sql_escaped = escape backslashes + single quotes"
    Ruby-->>OB: writes ALTER USER `user` PASSWORD '...' to $query_host
    OB->>OB: docker cp $query_host → container:$query_container
    OB->>IB: "stdin=existingPasswordFile, ARGV[1]=$query_container"
    IB->>IB: "cat >"$password_file" (existing password)"
    IB->>YDB: --password-file $password_file yql -f $query_container
    YDB-->>IB: success / error → $last_error
    IB-->>OB: rc
    OB-->>TS: result

    TS->>SB: "stdin=newPassword"
    SB->>SB: "cat >"$password_host" (new password)"
    SB->>RubySync: "ARGV[5]=$password_host (File.read UTF-8)"
    RubySync->>RubySync: update config YAML + write rootPasswordFile
    RubySync-->>SB: done
    SB-->>TS: result
```

<sub>Reviews (8): Last reviewed commit: ["test(local-ydb): assert password rotatio..."](https://github.com/astandrik/local-ydb-toolkit/commit/811ab7a0a33acad20274ef719e7fd233d8e8afb2) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=32624893)</sub>

### CODE REVIEW

File: packages/core/src/operations/auth-operations.ts [hunk 1]
```diff
@@ -250,16 +250,40 @@ export async function setRootPassword(
...
+        `File.write(ARGV[0], "ALTER USER ${ctx.profile.rootUser} PASSWORD '#{sql_escaped}';\\n")`
+      ].join("; ")),
+      "\"$query_host\""
+    ].join(" "),
     `rotate_with_password_file() {
   local file="$1"
   [ -f "$file" ] || return 1
-  cat "$file" | docker exec -i ${shellQuote(ctx.profile.staticContainer)} bash -lc ${shellQuote(`umask 077; cat >/tmp/root.password; /ydb -e grpc://localhost:${ctx.profile.ports.dynamicGrpc} -d ${shellQuote(ctx.profile.tenantPath)} --user ${shellQuote(ctx.profile.rootUser)} --password-file /tmp/root.password yql -s "ALTER USER ${ctx.profile.rootUser} PASSWORD '${escapedPassword}';"; rc=$?; rm -f /tmp/root.password; exit $rc`)} >"$last_error" 2>&1
+  query_container=$(docker exec ${shellQuote(ctx.profile.staticContainer)} mktemp /tmp/local-ydb-toolkit-password-rotate-XXXXXX.yql)
+  docker cp "$query_host" ${shellQuote(`${ctx.profile.staticContainer}:`)}"$query_container"
```
[the remark this answers is not in the document]
Comment (astandrik) [reply packages/core/src/operations/auth-operations.ts]: Addressed locally in the current branch. `rotate_with_password_file` now treats both `docker exec ... mktemp` and `docker cp` as explicit failure points, cleans up the container temp path on copy failure, and returns before invoking `yql -f` when the copy step fails. Reverified with focused tests, full `npm test`, `npm run typecheck`, `npm run build`, and a live rotate/restore check against the `ghcr261-auth` stack.

File: packages/core/src/operations/auth-operations.ts [hunk 2]
```diff
@@ -250,16 +250,40 @@ export async function setRootPassword(
...
+  set +e
+  cat "$file" | docker exec -i ${shellQuote(ctx.profile.staticContainer)} bash -lc ${shellQuote(`query_file="$1"; umask 077; cat >/tmp/root.password; trap 'rm -f /tmp/root.password "$query_file"' EXIT; /ydb -e grpc://localhost:${ctx.profile.ports.dynamicGrpc} -d ${shellQuote(ctx.profile.tenantPath)} --user ${shellQuote(ctx.profile.rootUser)} --password-file /tmp/root.password yql -f "$query_file"`)} _ "$query_container" >"$last_error" 2>&1
```
[the remark this answers is not in the document]
Comment (astandrik) [reply packages/core/src/operations/auth-operations.ts]: Confirmed and fixed locally. Runtime rotation now uses a per-invocation container-side `mktemp` password file instead of the shared `/tmp/root.password`, and I also switched the authenticated wait helper off the shared container path so the end-to-end rotation flow no longer races on that file. Reverified with focused tests, full `npm test`, `npm run typecheck`, `npm run build`, and a live rotate/restore check against `ghcr261-auth`.

File: packages/core/src/operations/auth-operations.ts [hunk 3]
```diff
@@ -250,16 +250,40 @@ export async function setRootPassword(
...
+    query_container=
+  fi
+}`,
+    "trap 'rm -f \"$candidate\" \"$last_error\" \"$query_host\"; cleanup_query_container' EXIT",
+    [
+      "ruby -e",
+      shellQuote([
+        "password = STDIN.read",
+        "password = password.sub(/\\n\\z/, \"\")",
+        "sql_escaped = password.gsub(\"'\", \"''\")",
```
[the remark this answers is not in the document]
Comment (astandrik) [reply packages/core/src/operations/auth-operations.ts]: Validated against the current code path and the live `ghcr261-auth` stack. `ShellCommandExecutor` does pass `spec.stdin` through unchanged, so trimming a trailing newline was incorrect. I removed that trim and made the tool reject passwords containing `\r` or `\n` up front in both the runtime guard and the MCP validation surface (`SetRootPasswordArgs` and JSON Schema), which keeps runtime rotation and host-side sync aligned and avoids transport ambiguity. Full tests/typecheck/build and live rotate/restore all passed after the change.

File: packages/core/src/operations/auth-operations.ts [hunk 4]
```diff
@@ -294,7 +320,7 @@ export async function setRootPassword(
     `target=$(${targetCommand})`,
     `docker exec ${shellQuote(ctx.profile.staticContainer)} cat "$target" > "$cfg_tmp"`,
     [
-      "ruby -ryaml -e",
+      "LOCAL_YDB_NEW_PASSWORD=\"$password\" ruby -ryaml -e",
```
[the remark this answers is not in the document]
Comment (astandrik) [reply packages/core/src/operations/auth-operations.ts]: Handled locally. Host-side auth sync no longer passes the new password through `LOCAL_YDB_NEW_PASSWORD`; it now captures the exact bytes from stdin into a local temp file and has Ruby read that file directly, which removes the process-environment exposure and keeps the persisted password aligned with the runtime rotation path. Reverified with focused tests, full `npm test`, `npm run typecheck`, `npm run build`, and a live rotate/restore check against `ghcr261-auth`.

File: packages/core/src/operations/auth-operations.ts [hunk 5]
```diff
@@ -250,16 +260,43 @@ export async function setRootPassword(
...
+  fi
+}`,
+    "trap 'rm -f \"$candidate\" \"$last_error\" \"$query_host\"; cleanup_query_container' EXIT",
+    [
+      "ruby -e",
+      shellQuote([
+        "password = STDIN.read",
+        "password = password.sub(/\\n\\z/, \"\")",
+        "sql_escaped = password.gsub(\"'\", \"''\")",
+        `File.write(ARGV[0], "ALTER USER ${ctx.profile.rootUser} PASSWORD '#{sql_escaped}';\\n")`
```
[the remark this answers is not in the document]
Comment (astandrik) [reply packages/core/src/operations/auth-operations.ts]: Addressed locally. The rotation query generator now receives `rootUser` as an argv value instead of interpolating it into the Ruby source, then emits it through a YQL identifier quoting helper before writing the `ALTER USER` query. Added a regression test with a hostile `rootUser` string to keep this path from regressing.

File: packages/mcp-server/src/tools/input-schemas.ts [hunk 6]
```diff
@@ -570,8 +570,10 @@ export function setRootPasswordSchema(): Tool["inputSchema"] {
       },
       password: {
         type: "string",
+        minLength: 1,
+        pattern: "^[^\\r\\n]+$",
```
[the remark this answers is not in the document]
Comment (astandrik) [reply packages/mcp-server/src/tools/input-schemas.ts]: Addressed locally. Added MCP-side coverage in `packages/mcp-server/test/tools.test.ts` for the `local_ydb_set_root_password` JSON Schema contract (`required`, `minLength`, and the CR/LF pattern) plus Zod rejection coverage for empty, newline, and carriage-return passwords. Also tightened the schema pattern to avoid the ECMAScript/JSON Schema `$` trailing-line-terminator edge case.

File: packages/core/test/operations.test.ts [hunk 7]
```diff
@@ -1322,6 +1326,61 @@ describe("mutating operations", () => {
...
+        }
+      }
+    }));
+
+    const response = await setRootPassword(ctx, { password: "line1\nline2" });
+
+    expect(response.executed).toBe(false);
+    expect(response.summary).toContain("does not support passwords containing carriage returns or newlines");
+    expect(response.plannedCommands).toEqual([]);
+  });
```
[the remark this answers is not in the document]
Comment (astandrik) [reply packages/core/test/operations.test.ts]: Addressed locally. Parameterized the root-password rejection test to exercise both `\n` and `\r`, so carriage-return rejection is covered independently from newline rejection.

File: packages/core/src/operations/auth-operations.ts [hunk 8]
```diff
@@ -61,7 +61,7 @@ function waitForAuthenticatedTenantStatusSpec(ctx: ToolkitContext) {
   }
 
   const withPassword = (innerCommand: string) => {
-    const script = `umask 077; cat >/tmp/root.password; ${innerCommand}; rc=$?; rm -f /tmp/root.password; exit $rc`;
+    const script = `password_file=$(mktemp /tmp/local-ydb-toolkit-root-password-XXXXXX); umask 077; cat >"$password_file"; trap 'rm -f "$password_file"' EXIT; ${innerCommand.replaceAll("/tmp/root.password", '"$password_file"')}`;
```
[the remark this answers is not in the document]
Comment (astandrik) [reply packages/core/src/operations/auth-operations.ts]: Addressed locally. Both container-side password-file paths now install their cleanup trap immediately after `mktemp` and before `cat >"$password_file"`, so an interruption during stdin copy no longer leaves an untrapped password temp file behind. Added regression assertions in the root-password planning test to verify the trap-before-write ordering for both runtime rotation and authenticated verification. Reverified with focused test, full `npm test`, `npm run typecheck`, and `npm run build`.

File: packages/core/src/operations/auth-operations.ts [hunk 9]
```diff
@@ -61,7 +61,7 @@ function waitForAuthenticatedTenantStatusSpec(ctx: ToolkitContext) {
   }
 
   const withPassword = (innerCommand: string) => {
-    const script = `umask 077; cat >/tmp/root.password; ${innerCommand}; rc=$?; rm -f /tmp/root.password; exit $rc`;
+    const script = `umask 077; password_file=$(mktemp /tmp/local-ydb-toolkit-root-password-XXXXXX); trap 'rm -f "$password_file"' EXIT; cat >"$password_file"; ${innerCommand.replaceAll("/tmp/root.password", '"$password_file"')}`;
```
[the remark this answers is not in the document]
Comment (astandrik) [reply packages/core/src/operations/auth-operations.ts]: Validated the current helper usage against the latest code. Both current callers still pass the intended /tmp/root.password placeholder, so behavior is correct as shipped. I kept this PR scoped to credential-handling fixes rather than reshaping the helper interface, and I am resolving this outdated thread on that basis.

File: packages/core/src/operations/auth-operations.ts [hunk 10]
```diff
@@ -307,7 +353,7 @@ export async function setRootPassword(
         "security[\"register_dynamic_node_allowed_sids\"] = allowed_sids",
         "root = Array(security[\"default_users\"]).find { |user| user[\"name\"] == \"root\" }",
         "raise \"root password not found in security_config.default_users\" unless root",
-        `password = ${rubyPasswordLiteral}`,
+        "password = File.binread(ARGV[5])",
```
[the remark this answers is not in the document]
Comment (astandrik) [reply packages/core/src/operations/auth-operations.ts]: Addressed locally. Host-side auth sync now reads the temp password file as UTF-8 with File.read(..., mode: "r:UTF-8") before YAML serialization, so non-ASCII passwords stay a normal scalar string instead of !!binary. Reverified with a Ruby YAML probe plus focused tests, full npm test, typecheck, and build.

File: packages/core/src/operations/auth-operations.ts [hunk 11]
```diff
@@ -250,16 +260,48 @@ export async function setRootPassword(
...
+    docker exec ${shellQuote(ctx.profile.staticContainer)} rm -f "$query_container" >/dev/null 2>&1 || true
+    query_container=
+  fi
+}`,
+    "trap 'rm -f \"$candidate\" \"$last_error\" \"$query_host\"; cleanup_query_container' EXIT",
+    [
+      "ruby -e",
+      shellQuote([
+        "def yql_identifier(value)",
+        "escaped = value.gsub(\"\\\\\") { \"\\\\\\\\\" }.gsub(\"`\") { \"\\\\`\" }",
```
[the remark this answers is not in the document]
Comment (astandrik) [reply packages/core/src/operations/auth-operations.ts:280]: Validated against the YDB lexer docs and a live read-only query on ghcr261-auth: backslash-escaped backticks inside a delimited identifier parse correctly and produce the expected identifier name, so I am leaving the current identifier escaping as is. No code change for this one.

File: packages/core/src/operations/auth-operations.ts [hunk 12]
```diff
@@ -279,18 +321,22 @@ export async function setRootPassword(
     "cat \"$last_error\" >&2",
     "exit 1"
   ].join("\n"), {
+    stdin: password,
```
[the remark this answers is not in the document]
Comment (astandrik) [reply packages/core/src/operations/auth-operations.ts:324]: Addressed locally in commit 811ab7a. The root-password rotation test now captures the generated CommandSpec objects and asserts that both the runtime rotation command and the host-side sync command carry the supplied password through spec.stdin, while the displayed planned commands still redact the password. Verified with focused tests, full npm test/typecheck/build, and a live rotate/restore smoke on ydb-local-ghcr261.
